src/track/globaltrackcache.cpp

```diff
@@ -139,41 +154,111 @@ GlobalTrackCache* volatile GlobalTrackCache::s_pInstance = nullptr;
 
 //static
 void GlobalTrackCache::createInstance(GlobalTrackCacheEvictor* pEvictor) {
-    DEBUG_ASSERT(s_pInstance == nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(!s_pInstance);
     s_pInstance = new GlobalTrackCache(pEvictor);
 }
 
 //static
 void GlobalTrackCache::destroyInstance() {
-    DEBUG_ASSERT(s_pInstance != nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(s_pInstance);
     GlobalTrackCache* pInstance = s_pInstance;
     s_pInstance = nullptr;
     delete pInstance;
 }
 
 //static
 void GlobalTrackCache::deleter(Track* pTrack) {
-    if (s_pInstance != nullptr) {
-        s_pInstance->evict(pTrack);
+    DEBUG_ASSERT(pTrack);
+    // Any access to pTrack is forbidden!! Due to race condition
+    // this pointer might already have been deleted! Only the
+    // cache knows this.
```

Comment:
Is this still true?
Why suffers a smart pointer a race condition? Or is the deleter also used other points then the smart pointer delete hook?

Reply:
Yes, of course! The deleter is invoked before the cache is locked. The reference counter might drop to 0 multiple times, before the track is actually deleted. Due to race conditions a previous decrease and invocation might be outpaced by a following decrease which already triggered the deletion of the track. I didn't take this into account first. Without locks almost everything is possible!!!